src/version_req.rs

```diff
@@ -407,6 +408,315 @@ impl VersionReq {
+    ///
+    /// Panics if `self` and `other` have different compatibilities.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use semver::VersionReq;
+    /// use semver::Version;
+    ///
+    /// let version_110 = Version { major: 1, minor: 1, patch: 0, pre: vec![], build: vec![] };
+    /// let version_111 = Version { major: 1, minor: 1, patch: 1, pre: vec![], build: vec![] };
+    /// let match_110 = VersionReq::parse("=1.1.0").unwrap();
+    /// let match_111 = VersionReq::parse(">1.1.0").unwrap();
+    /// let either = match_110.union(&match_111);
+    ///
+    /// assert!(either.matches(&version_110));
+    /// assert!(either.matches(&version_111));
+    /// ```
+    pub fn union(mut self, other: &VersionReq) -> VersionReq {
+        assert_eq!(self.compat, other.compat);
+        self.ranges.extend(other.ranges.iter().cloned());
+        self.simplify();
+        self
+    }
+
+    /// Return a `VersionReq` that matches any version that matches the current requirements _and_
+    /// those expressed in `other`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `self` and `other` have different compatibilities.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use semver::VersionReq;
+    /// use semver::Version;
+    ///
+    /// let version_110 = Version { major: 1, minor: 1, patch: 0, pre: vec![], build: vec![] };
+    /// let version_111 = Version { major: 1, minor: 1, patch: 1, pre: vec![], build: vec![] };
+    /// let version_120 = Version { major: 1, minor: 2, patch: 0, pre: vec![], build: vec![] };
+    /// let gt_110 = VersionReq::parse(">1.1.0").unwrap();
+    /// let lt_120 = VersionReq::parse("<1.2.0").unwrap();
+    /// let both = gt_110.intersection(&lt_120);
+    ///
+    /// assert!(both.matches(&version_111));
+    /// assert!(!both.matches(&version_110));
+    /// assert!(!both.matches(&version_120));
+    /// ```
+    pub fn intersection(mut self, other: &VersionReq) -> VersionReq {
+        assert_eq!(self.compat, other.compat);
+
+        let intersection = Vec::with_capacity(self.ranges.len() * other.ranges.len());
+
+        // We allow any range that matches a predicate from self and a predicate from other.
+        // It's not important _which_ predicate from each it matters.
+        for p1 in mem::replace(&mut self.ranges, intersection) {
+            assert_eq!(self.compat, p1.compat);
+            for p2 in other.ranges.iter() {
+                assert_eq!(self.compat, p2.compat);
+
+                // Allow anything that matches p1 & p2
+                self.ranges.push(Range {
+                    compat: self.compat,
+                    predicates: p1
+                        .predicates
+                        .iter()
+                        .chain(p2.predicates.iter())
+                        .cloned()
+                        .collect(),
+                });
+            }
+        }
+        self.simplify();
+
+        self
+    }
+
+    fn simplify(&mut self) {
```

Comment:
this is not pub.

Reply:
Hmm, I don't think it should be actually. I think `simplify` should be performed internally when it makes sense to do so (like after a call to `intersection`). It's not clear there's value in users being able to call it randomly.